model/RdsResultStorage.php

```diff
@@ -273,263 +187,260 @@ public function getVariables($callId)
-            $limit = (isset($options['limit']))?$options['limit']:1000;
-            $sql = $this->getPersistence()->getPlatForm()->limitStatement($sql, $limit, $offset);
-        }
-        $results = $this->getPersistence()->query($sql, $params);
-        foreach ($results as $value) {
-            $returnValue[] = array(
-                "deliveryResultIdentifier" => $value[self::RESULTS_TABLE_ID],
-                "testTakerIdentifier" => $value[self::TEST_TAKER_COLUMN],
-                "deliveryIdentifier" => $value[self::DELIVERY_COLUMN]
-            );
+        $returnValue = [];
+        foreach ($qb->execute()->fetchAll() as $value) {
+            $returnValue[] = [
+                self::FIELD_DELIVERY_RESULT => $value[self::RESULTS_TABLE_ID],
+                self::FIELD_TEST_TAKER => $value[self::TEST_TAKER_COLUMN],
+                self::FIELD_DELIVERY => $value[self::DELIVERY_COLUMN],
+            ];
         }
+
         return $returnValue;
+    }
+
+    /**
+     * Generates and sanitize ORDER BY field.
+     *
+     * @param array $options
+     *
+     * @return string
+     */
+    protected function getOrderField(array $options)
+    {
+        $allowedOrderFields = [self::DELIVERY_COLUMN, self::TEST_TAKER_COLUMN, self::RESULTS_TABLE_ID];
+
+        if (isset($options['order']) && in_array($options['order'], $allowedOrderFields)) {
+            return $options['order'];
+        }
 
+        return self::RESULTS_TABLE_ID;
     }
 
-    public function countResultByDelivery($delivery){
-        $sql = 'SELECT COUNT(*) FROM ' . self::RESULTS_TABLENAME;
-        $params = array();
+    /**
+     * Generates and sanitize ORDER BY direction.
+     *
+     * @param array $options
+     *
+     * @return string
+     */
+    protected function getOrderDirection(array $options)
+    {
+        $allowedOrderDirections = ['ASC', 'DESC'];
+
+        if (isset($options['orderdir']) && in_array(strtoupper($options['orderdir']), $allowedOrderDirections)) {
+            return $options['orderdir'];
+        }
+
+        return 'ASC';
+    }
 
+    public function countResultByDelivery($delivery)
+    {
+        if (!is_array($delivery)) {
+            $delivery = [$delivery];
+        }
+        $qb = $this->getQueryBuilder()
+            ->select('COUNT(*)')
+            ->from(self::RESULTS_TABLENAME);
 
         if (count($delivery) > 0) {
-            $sql .= ' WHERE ';
-            $inQuery = implode(',', array_fill(0, count($delivery), '?'));
-            $sql .= self::DELIVERY_COLUMN . ' IN (' . $inQuery . ')';
-            $params = array_merge($params, $delivery);
+            $qb
+                ->andWhere(self::DELIVERY_COLUMN .' IN (:delivery)')
+                ->setParameter('delivery', $delivery, Connection::PARAM_STR_ARRAY);
         }
 
-        return $this->getPersistence()->query($sql, $params)->fetchColumn();
+        return $qb->execute()->fetchColumn();
```

Comment:
Is it safe to execute the `COUNT` even without `WHERE` clause? (when delivery array is empty)

Reply:
What do you mean by safe? It then turns out to a count(*) on a table. It's covered in the test (line 131).
In fact, there is an impact on Spanner performance though, since Spanner doesn't store statistics on the tables (because they're sharded), and for big tables that would turn into long query execution. But this is on the side of optimisation.